### ISSUE TITLE
Offer Python Latest (nightly master trunk build)

### ISSUE THREAD

**Issue by cooperlees** [actions/setup-python#31]
One of the features I love most about TravisCI is the ability to build my projects CI with top of Trunk Python (master branch). This allows me to know about deprecations etc. coming very early and have time to work on the fixes sooner.

Can actions look at supporting building a Trunk version nightly? Here is an example usage with Travis: https://github.com/facebookincubator/ptr/blob/master/.travis.yml#L16

Haven't dug in deep, but would also love to have the ability for this build to be able to fail and not flag CI failure, like the Travis config above does.

Thanks.

**Comment by hugovk** [comment]
cf. https://docs.travis-ci.com/user/languages/python/#nightly-build-support

and https://docs.travis-ci.com/user/languages/python/#development-releases-support

**Comment by isidentical** [comment]
@madhurig any news?

**Comment by cclauss** [comment]
I would not call this Python __Latest__ but would instead call it Python __Nightly__.

**Comment by hugovk** [comment]
Is this a duplicate of https://github.com/actions/setup-python/issues/20?

It has a workaround to use deadsnakes: https://github.com/actions/setup-python/issues/20#issuecomment-596221879.

**Comment by asottile** [comment]
there's now a slightly better workaround using [deadsnakes/action](https://github.com/deadsnakes/action)

**Comment by DeepSpace2** [comment]
This should really be offered "natively" by setup-python.

**Comment by rouilj** [comment]
Agreed, having 3.12-alpha available would be a big win. Especially since xapian fails to build and a
number of other deprecated features like cgi are removed IIRC in 3.12 and they need testing.

**Comment by kelson42** [comment]
`3.12` [has been released yesterday](https://docs.python.org/3.12/whatsnew/3.12.html)

**Comment by pquentin** [comment]
> `3.12` [has been released yesterday](https://docs.python.org/3.12/whatsnew/3.12.html)

This is only alpha 6, see https://peps.python.org/pep-0693/ for the full 3.12 release schedule.

**Comment by kelson42** [comment]
@pquentin ooops. Thx

**Comment by aparnajyothi-y** [comment]
Hello @cooperlees, Thank you for your suggestion regarding nightly builds for Python versions. While we recognize the value of having the latest updates, providing nightly builds directly from the source code on a daily basis would be challenging. We appreciate your understanding and continued support, will keep you informed of any future developments or alternatives, and encourage you to share any future suggestions.

**Comment by AraHaan** [comment]
For nightly builds could have the action run ``git clone https://github.com/python/cpython.git --branch main`` and then run the required steps to build cpython from source before using it like normal. I do something similar to this locally.

**Comment by cclauss** [comment]
Also to consider:
* astral-sh/setup-uv#124

**Comment by v-aparnajyothi-y** [comment]
Hello Everyone, Thanks for the suggestion. After consideration, we’ve decided not to support this in setup-python as our goal is to provide a stable and predictable experience using officially supported Python versions. Nightly builds are inherently unstable, and would add significant maintenance burden and risk of confusion for users. For advanced needs, we recommend using custom installation steps. Appreciate your feedback!

**Comment by v-aparnajyothi-y** [comment]
Hello @cooperlees, Please let us know if you need any clarifications or any concerns on the above :)

**Comment by cooperlees** [comment]
> Nightly builds are inherently unstable, and would add significant maintenance burden and risk of confusion for users

I feel people who want to do this are power users generally, and understand the risks. Feel free to flag these options with warning that the python environment for nightly builds is best effort and unsupported. We'd all accept that.

But it seems we'll just have to use own custom/other actions to do this. My goal here was to just make it easier if it was not to much work for setup-python. While my only concern is I'm disappointed, but respect the teams decision.

**Comment by v-aparnajyothi-y** [comment]
Hello @cooperlees,Thank you for your understanding, you're absolutely right that power users often recognize and accept the risks associated with using nightly builds. Your suggestion to include a clear disclaimer about their instability and lack of support is a thoughtful one, and we’ll certainly keep it in mind if we revisit this in the future.

For now, in order to uphold the stability guarantees that setup-python aims to provide, we’ll continue to recommend using custom installation steps for nightly or experimental versions. That said, we’re always open to reassessing decisions based on community feedback and evolving ecosystem needs, so please do keep sharing your use cases and suggestions.

Thanks again for the constructive engagement and feel free to reach out if you need any further support.

**Comment by nedbat** [comment]
I had been using deadsnakes for nightly builds, but it wasn't always available in a timely way and I couldn't understand their build process.  I ended up using pyenv in my own action workflow:  https://github.com/nedbat/coveragepy/blob/master/.github/workflows/python-nightly.yml

But it would still be great for a more central implementation that others can rely on.

**Comment by v-aparnajyothi-y** [comment]
Hi @nedbat, Thanks for the input! To maintain the stability guarantees that setup-python aims to provide, we currently recommend using custom installation steps for nightly or experimental versions. That said, we’re always open to revisiting this based on community feedback and evolving needs so please keep the suggestions coming.

We truly appreciate your continued engagement and support. Please feel free to reach out if you need any further support!

**Comment by hauntsaninja** [comment]
Just posting to provide some more community feedback that this feature would be useful

**Comment by dopry** [comment]
@aparnajyothi-y I think you're short changing your users. The users of CI are developers. Github is not a 'hosting' platform. It is a development platform. As a developer and project lead I need to consistently test against the newest version of my platform language to catch changes that have major impacts on projects in advance, so we can address it in our code base or raise issues upstream to ensure a well tested stable release. It seems very condescending to protect developers from doing their jobs in a way that adversely effects the entire ecosystem.

**Comment by toonarmycaptain** [comment]
Speaking for myself and the various teams I'm a part of, _stability_ is exactly what I would hope setup-python would provide with nightly/dev/beta versions of python. Not that those version of python themselves will be stable, but a consistent installation mechanism from version to version, whether it's 3.12 or 3.16-alpha is part of the value that this project provides. The alternative can be fragile hacks to get various versions which isn't fun.
